### PR TITLE
#163188330 remove user role entity

### PR DESCRIPTION
**Pull request by dnuwa**
#### What does this PR do?
The PR Refactors code to remove/delete the user_roll entity and instead use an association table to manage the many to many relationships between the users' table and the role table

#### Description of Task to be completed?
this is to replace the user_role entity(model and scheme) with an association table.

#### How to manually test it.
Make migrations to update the user_role table
Run the server http://127.0.0.1:5000/mrm
Run the following query for assigning the role to a user 
```
mutation{
  createUserRole(userId: 10, roleId: 1){
    userRole{
      email
    }
  }
}
```
Run the following query to edit the user's role
```
mutation{
    changeUserRole(email:"danie.nuwagaba@andela.com", roleId:2){
        user{
            name
            roles{
                role
            }
        }
    }
}
```
#### screenshots
image

image

#### What are the relevant pivotal tracker stories?
[163188330](https://www.pivotaltracker.com/story/show/163188330)

### CODE REVIEW

**Comment by dnuwa** [comment]
> I think we can have a better description on how you accomplished the task on your PR and please fix the tests and rebase so that we can all review this PR.

Thanx for this review @michael-basweti. I am working on this.

File: tests/test_user_roles/test_query_user_role.py [hunk 1]
```diff
@@ -34,36 +35,37 @@ def test_query_users_role(self):
...
 
-    def test_query_users_role_by_role(self):
-        user = User(email='mrm@andela.com', location="Lagos",
-                    name="test test",
-                    picture="www.andela.com/testuser")
-        user.save()
-        user_role = UsersRole(user_id=user.id, role_id=1)
-        user_role.save()
-        db_session().commit()
+    # def test_query_users_role_by_role(self):
```
Comment (C3real-kill3r) [inline tests/test_user_roles/test_query_user_role.py]: hello, I'd love to know why these tests commented out?

Comment (caveinn) [reply tests/test_user_roles/test_query_user_role.py]: +1

File: alembic/versions/0d8aed449d2f_migration_message.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    pass
+    # ### end Alembic commands ###
+
+
+def downgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    pass
+    # ### end Alembic commands ###
```
Comment (cwizard2011) [inline alembic/versions/0d8aed449d2f_migration_message.py]: You should have just one migration file, the one that defines the association table, kindly remove other unrelated alembic versions

File: alembic/versions/2c4af13eef71_migration_message.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    op.drop_table('wings')
+    op.drop_table('floors')
+    op.drop_table('blocks')
+    op.drop_table('users_roles')
+    op.drop_table('offices')
+    op.drop_table('notifications')
+    op.drop_table('users')
+    op.drop_table('roles')
+    op.drop_table('questions')
+    op.drop_table('locations')
```
Comment (cwizard2011) [inline alembic/versions/2c4af13eef71_migration_message.py]: This alembic file is not necessary, please remove this

File: alembic/versions/d7d76e7557b9_user_role_table.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    op.drop_table('floors')
+    op.drop_table('blocks')
+    op.drop_table('users_roles')
+    op.drop_table('offices')
+    op.drop_table('notifications')
+    op.drop_table('users')
+    op.drop_table('roles')
+    op.drop_table('questions')
+    op.drop_table('locations')
+    # ### end Alembic commands ###
```
Comment (cwizard2011) [inline alembic/versions/d7d76e7557b9_user_role_table.py]: remove this

File: tests/test_user_roles/test_query_user_role.py [hunk 5]
```diff
@@ -1,70 +1,21 @@
...
 
 import sys
 import os
+import tests.base as base
 sys.path.append(os.getcwd())
 
 
 class TestQueryUserRole(BaseTestCase):
 
-    def test_query_users_role(self):
```
Comment (cwizard2011) [inline tests/test_user_roles/test_query_user_role.py]: Fix this test

File: alembic/versions/2c4af13eef71_migration_message.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
+"""Migration message
```
Comment (Bonifase) [inline alembic/versions/2c4af13eef71_migration_message.py]: Since you are modifying just role and user models, it is not necessary to run fresh migration for the entire database hence these migration files should not contain migration for the entire database. Also, kindly have only one migration file in your PR. Please work with the backend team to guide you through this. Please review other team member's PRs to avoid having the same slips in your PR.

**Comment by kampkelly** [comment]
Good progress. Please implement @cwizard2011 and @Bonifase feedback on alembic files and the test.

File: alembic/versions/0d8aed449d2f_migration_message.py (hunk 2 above)
Comment (dnuwa) [reply alembic/versions/0d8aed449d2f_migration_message.py]: This has been worked on @cwizard2011. thanx for pointing it out.

File: alembic/versions/2c4af13eef71_migration_message.py (hunk 3 above)
Comment (dnuwa) [reply alembic/versions/2c4af13eef71_migration_message.py]: this has been removed

File: alembic/versions/d7d76e7557b9_user_role_table.py (hunk 4 above)
Comment (dnuwa) [reply alembic/versions/d7d76e7557b9_user_role_table.py]: done

File: alembic/versions/2c4af13eef71_migration_message.py (hunk 6 above)
Comment (dnuwa) [reply alembic/versions/2c4af13eef71_migration_message.py]: Thank you for pointing these issues out @Bonifase  I will endeavor to do as advised.

File: tests/test_user_roles/test_query_user_role.py [hunk 7]
```diff
@@ -1,70 +1,15 @@
...
+    change_unavailable_user_role_mutation_response)
 import sys
 import os
 sys.path.append(os.getcwd())
 
 
 class TestQueryUserRole(BaseTestCase):
 
-    def test_query_users_role(self):
-        """
```
[the remark this answers is not in the document]
Comment (caveinn) [reply tests/test_user_roles/test_query_user_role.py]: +1

[the remark this answers is not in the document]
Comment (dnuwa) [reply tests/test_user_roles/test_query_user_role.py]: I am currently implementing this feedback

File: tests/test_user/test_delete_user.py [hunk 8]
```diff
@@ -27,30 +27,14 @@ def test_deleteuser_when_admin(self):
...
-        user_role.save()
+        user.roles.append(role)
 
         CommonTestCases.admin_token_assert_equal(
             self,
             delete_user_2,
             expected_query_after_delete
         )
 
-    def test_user_delete_admin(self):
```
[the remark this answers is not in the document]
Comment (dnuwa) [reply tests/test_user/test_delete_user.py]: I am currently working on the same to ensure that the test passes.

File: api/user/schema.py [hunk 9]
```diff
@@ -112,11 +113,46 @@ def mutate(self, info, email, **kwargs):
...
+        if not exact_user:
+            raise GraphQLError('user not found')
+
+        role = Role.get_query(info)
+        exact_role = role.filter_by(id=kwargs['role_id']).first()
+
+        if not exact_role:
+            raise GraphQLError('Role id does not exist')
+
+        if len(exact_user.roles) > 0:
```
Comment (Bonifase) [inline api/user/schema.py]: I wouldn't recommend this method of check. Use a simpler method by simply checking if the list in non-empty without necessarily using the len method and the greater than symbol.

**Comment by RonKbS** [comment]
Nice work @dnuwa,

I have pulled your work and notice you still have the user_role folder in place with a model file in it.
Is it still required?

**Comment by dnuwa** [comment]
> Nice work @dnuwa,
> 
> I have pulled your work and notice you still have the user_role folder in place with a model file in it.
> Is it still required?

no, it is no longer required and I have made sure to delete it.

File: api/user/schema.py [hunk 10]
```diff
@@ -112,11 +113,46 @@ def mutate(self, info, email, **kwargs):
...
-        user_role = UsersRole.query.filter_by(user_id=exact_user.id).first()
-        new_role = kwargs.pop('role_id')
-        user_role.role_id = new_role
-        with SaveContextManager(user_role, 'Role id', new_role):
-            return ChangeUserRole(user=exact_user)
+
+        if not exact_user.roles:
+            raise GraphQLError('user has no role')
+
+        new_role_obj = RoleModel.query.filter_by(id=kwargs['role_id']).first()
```
Comment (Bonifase) [inline api/user/schema.py]: What does 'new_role_obj' is supposed to represent here? Avoid abbreviations in naming your variables.

File: tests/test_user_roles/test_query_user_role.py (hunk 1 above)
Comment (dnuwa) [reply tests/test_user_roles/test_query_user_role.py]: This test was uncommented and refactored to pass. thanx @C3real-kill3r for pointing it out.

File: api/user/schema.py (hunk 10 above)
Comment (dnuwa) [reply api/user/schema.py]: noted. I have changed the variable to `new_role` which is meant to hold the new role queried from the `roles` table

File: tests/test_user/test_delete_user.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dnuwa) [reply tests/test_user/test_delete_user.py]: This was fixed @cwizard2011

File: api/user/schema.py [hunk 11]
```diff
@@ -112,11 +113,46 @@ def mutate(self, info, email, **kwargs):
...
-            return ChangeUserRole(user=exact_user)
+
+        if not exact_user.roles:
+            raise GraphQLError('user has no role')
+
+        new_role = RoleModel.query.filter_by(id=kwargs['role_id']).first()
+        if not new_role:
+            raise GraphQLError('invalid role id')
+
+        exact_user.roles[0] = new_role
```
Comment (joshuaocero) [inline api/user/schema.py]: @dnuwa any reason why here you update the user's role using the index while in the CreateUserRole mutation you use the append method? Try and be more consistent.

File: alembic/env.py [hunk 12]
```diff
@@ -89,4 +88,4 @@ def run_migrations_online():
 if context.is_offline_mode():
     run_migrations_offline()
 else:
-    run_migrations_online()
\ No newline at end of file
+    run_migrations_online()
```
Comment (joshuaocero) [inline alembic/env.py]: @dnuwa this change is unnecessary

File: helpers/auth/authentication.py [hunk 13]
```diff
@@ -82,8 +81,7 @@ def save_user(self):
                 try:
                     user_data = User(email=email, name=name, picture=picture)
                     user_data.save()
-                    user_role = UsersRole(
-                        user_id=user_data.id, role_id=role.id)
+                    user_role = user_data.roles.append(role)
                     user_role.save()
```
Comment (joshuaocero) [inline helpers/auth/authentication.py]: @dnuwa I think variable naming in these lines of code can cause a bit of confusion, please refactor. There is also a bit of an issue with the logical flow here; why are you saving the user_data object twice for example?

File: helpers/auth/authentication.py [hunk 14]
```diff
@@ -128,17 +126,20 @@ def wrapper(*args, **kwargs):
...
+                        if role.role in expected_args:
+                            return func(*args, **kwargs)
+                        else:
+                            message = (
+                                'You are not authorized to perform this action')
+                            status = 401
+                            handle_http_error(message, status, expected_args)
                     else:
-                        message = (
+                        raise GraphQLError(
```
Comment (joshuaocero) [inline helpers/auth/authentication.py]: @dnuwa why did you remove the http error handler here. I believe it is meant to help you raise these such GraphQL errors.

File: api/user/schema.py (hunk 11 above)
Comment (dnuwa) [reply api/user/schema.py]: `exact_user.roles` returns a list. so what I am trying to achieve here is to replace the value returned at index 0 with a new value in `new_role`. That's why the implementation here is different. This mutation takes care of an instance where the user needs to update a role.

File: alembic/env.py (hunk 12 above)
Comment (dnuwa) [reply alembic/env.py]: noted

File: api/user/schema.py [hunk 15]
```diff
@@ -112,11 +113,46 @@ def mutate(self, info, email, **kwargs):
...
+        role = Role.get_query(info)
+        exact_role = role.filter_by(id=kwargs['role_id']).first()
+
+        if not exact_role:
+            raise GraphQLError('Role id does not exist')
+
+        if exact_user.roles:
+            raise GraphQLError('user already has a role')
+
+        exact_user.roles.append(exact_role)
```
Comment (dnuwa) [inline api/user/schema.py]: Context: A user has no role when created. so, the append is used because the user to which a role is being appended has none at the moment.

File: templates/index.html [hunk 16]
```diff
@@ -1,7 +1,7 @@
 <!DOCTYPE html>
 <html lang="en">
 <head>
-    <meta name="google-site-verification" content="VBSIFp1OyI1gpnRTREBEjNBj_Wat50wzwzVTrG1YuHY" />
+    <meta name="google-site-verification" content="7nOT58tx-TbWUt6dokG5Yj1Chca3j5-qX5qQ4dTGVew" />
```
Comment (joshuaocero) [inline templates/index.html]: @dnuwa this should be raised in another PR.

Comment (dnuwa) [reply templates/index.html]: Noted

File: api/user/schema.py [hunk 17]
```diff
@@ -118,11 +119,46 @@ def mutate(self, info, email, **kwargs):
         exact_user = active_user.filter(UserModel.email == email).first()
         if not exact_user:
             raise GraphQLError("User not found")
-        user_role = UsersRole.query.filter_by(user_id=exact_user.id).first()
-        new_role = kwargs.pop('role_id')
-        user_role.role_id = new_role
-        with SaveContextManager(user_role, 'Role id', new_role):
-            return ChangeUserRole(user=exact_user)
+
+        if not exact_user.roles:
```
Comment (bevkololi) [inline api/user/schema.py]: Has this line been tested? Might be the reason the coverage is failing.

File: api/user/schema.py [hunk 18]
```diff
@@ -118,11 +119,46 @@ def mutate(self, info, email, **kwargs):
...
+            raise GraphQLError('user has no role')
+
+        new_role = RoleModel.query.filter_by(id=kwargs['role_id']).first()
+        if not new_role:
+            raise GraphQLError('invalid role id')
```
Comment (bevkololi) [inline api/user/schema.py]: This one too ...

Comment (dnuwa) [reply api/user/schema.py]: @bevkololi I have added a test to this.

File: api/user/schema.py (hunk 17 above)
Comment (dnuwa) [reply api/user/schema.py]: @bevkololi  thanx for pointing this out
